[Issue 11312] Data test run results always show failures

vglocus opened the pull request:
Resolves #11312 

### Problem

Previously, a RunResult for a data test always showed 0 failures for a passing test, no matter what the DataTestResult returned. For example a test that will error or warn if '> 10' and the result is 4 failures, the RunResult returns 0 failures, because it is passing.

### Solution

For data tests, always set failures to the actual failure count, even if technically not a failure.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me.
- [ ] I have run this code in development, and it appears to resolve the stated issue.
- [x] This PR includes tests, or tests are not required or relevant for this PR.
- [x] This PR has no interface changes (e.g., macros, CLI, logs, JSON artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX.
- [ ] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions.

jtcohen6 commented:
@vglocus Thanks for (re)opening this PR! I agree this is a good change. As @dbeatty10 explained in https://github.com/dbt-labs/dbt-core/issues/9808#issuecomment-2034482122, it is highly unlikely to represent a behavior change for anyone who's currently relying on the (IMO surprising & incorrect) `failures=0` to represent "success" for tests that return nonzero failing records that are less than the configured warn/error threshold.

Could you please update some of the functional tests [here](https://github.com/dbt-labs/dbt-core/blob/main/tests/functional/schema_tests/test_schema_v2_tests.py), so we can ensure this works going forward? (You can look https://github.com/dbt-labs/dbt-core/pull/9657 for inspiration; I'd recommend giving credit by adding @tbog357 as a co-contributor in the changelog entry.)

vglocus committed:
fix: preserve failure count in passing data test results

A RunResult for a data test always reported 0 failures when passing,
regardless of the actual DataTestResult. For example, a test
configured to error if '> 10' with 4 failures would return 0 in the
RunResult.

Fixes #11312

vglocus committed:
test: update functional tests for preserved failure count

Update assertTestPassed to allow failures >= 0 on passing tests,
and adjust expected failure sums in threshold-based test classes.

Fixes #11312

vglocus committed:
refactor: reduce test_failures_always_preserved argument count

Address CodeScene advisory by removing redundant parameters and
extracting the warn-severity case into a separate test method.

vglocus committed:
refactor: extract shared helper to reduce test duplication

Move common mock setup and test execution into _run_build_test
helper to address CodeScene code duplication advisory.

vglocus committed:
refactor: reduce _run_build_test to 3 parameters

Split result construction into _make_result helper so that
_run_build_test stays under the 4-argument advisory limit.

vglocus commented:
@jtcohen6 Would you mind taking another look at this one?

sambloom92 commented:
Hi @vglocus thanks for worKing on this PR, I would really like to see this get merged in, but could I suggest an improvement?

It would be really helpful if the `RunResult.message` attribute were also populated whenever there are test failures, regardless of the test status. In particular the [elementary](https://docs.elementary-data.com/data-tests/dbt/dbt-package) test report displays the result message on all tests, regardless of status, which would be really useful if it weren't for the fact that it's currently always empty for passed tests.

I was thinking something like this:
```
    @staticmethod
    def build_test_run_result(test: TestNode, result: TestResultData) -> RunResult:
        severity = test.config.severity.upper()
        thread_id = threading.current_thread().name
        num_errors = pluralize(result.failures, "result")
        failures = result.failures
        status = TestStatus.Pass
        message = None
        flags = get_flags()
        warn_error = flags.WARN_ERROR or flags.WARN_ERROR_OPTIONS.includes(
            LogTestResult.__name__
        )
        if severity == "ERROR" and result.should_error:
            status = TestStatus.Fail
            message = f"Got {num_errors}, configured to fail if {test.config.error_if}"
        elif result.should_warn:
            status = TestStatus.Fail if warn_error else TestStatus.Warn
            action = "fail" if warn_error else "warn"
            message = f"Got {num_errors}, configured to {action} if {test.config.warning_if}"
        elif failures:
            action = "fail" if severity == "ERROR" else "warn"
            threshold = (
                test.config.error_if if severity == "ERROR" else test.config.warning_if
            )
            message = f"Got {num_errors}, configured to {action} if {threshold}"
        run_result = RunResult(
            node=test,
            status=status,
            timing=[],
            thread_id=thread_id,
            execution_time=0,
            message=message,
            adapter_response=result.adapter_response,
            failures=failures,
            batch_results=None,
        )
        return run_result
```

Could I raise a PR into your branch with this change? I would be happy to help with the testing too.
